Re-implemented Validator Factory Pattern

hmarp opened the pull request:
Re-implemented Validator Factory Pattern

hmarp committed:
Re-implemented Validator Factory Pattern.

IMarples commented on PasswordValidator/Controllers/PasswordValidationController.cs:
```diff
@@ -4,14 +4,13 @@ namespace PasswordValidator.Controllers
     {
-        readonly SimpleValidatorFactory _SimpleValidatorFactory = new SimpleValidatorFactory();
-        readonly AdvancedValidatorFactory _AdvancedValidatorFactory = new AdvancedValidatorFactory();
+        readonly ValidatorFactory _validatorFactory = new ValidatorFactory();
```
Need to look at Dependency Injection - https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-6.0

hmarp committed:
Modified PasswordValidationController to take IValidatorFactory as input dependency.
Injected PasswordValidatorFactory into Controller.
Added IValidatorFactory Interface and made class PasswordValidatorFactory implement it.

IMarples commented on PasswordValidator_UnitTests/AdvancedValidator_UnitTests.cs:
```diff
@@ -6,14 +6,14 @@ namespace PasswordValidator_UnitTests
     internal class AdvancedValidator_UnitTests
     {
-        AdvancedValidatorFactory _factory = new AdvancedValidatorFactory();
+        IValidatorFactory _factory = new PasswordValidatorFactory();
```
In AdvancedValidator_UnitTests - you should just create an instance of the concrete class "AdvancedValidator" - you should have no reference to the factory. These tests should NOT know the factory exists.

IMarples commented on PasswordValidator/Controllers/PasswordValidationController.cs:
```diff
@@ -4,14 +4,18 @@ namespace PasswordValidator.Controllers
     {
-        readonly SimpleValidatorFactory _SimpleValidatorFactory = new SimpleValidatorFactory();
-        readonly AdvancedValidatorFactory _AdvancedValidatorFactory = new AdvancedValidatorFactory();
+        readonly IValidatorFactory _validatorFactory;
```
Declare variables as var - https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/var

hmarp replied:
No longer applies here?

IMarples replied:
Sorry the var thing is in methods/functions - however, variables need declaring correctly for their SCOPE - these are.. private readonly variables and that is important because they can only be initiated in the constructor, hence only by DI

IMarples commented on PasswordValidator/Program.cs:
```diff
@@ -1,8 +1,12 @@
 // Add services to the container.
 
 builder.Services.AddControllers();
+builder.Services.AddScoped<IValidatorFactory, ValidatorFactory>();
```
Good

IMarples commented on PasswordValidator_UnitTests/AdvancedValidator_UnitTests.cs:
```diff
@@ -26,7 +26,7 @@ public void IsValidAdvancedPassword_ValidPassword_ShouldReturnTrue(string advanc
+            IPasswordValidator validator = _factory.GetPasswordValidator("Advanced");
 
             bool result = validator.Validate(advancedPassword);
 
```
How would you test for an exception being thrown? Can you force it to throw an exception?

IMarples commented on PasswordValidator_UnitTests/SimpleValidator_UnitTests.cs:
```diff
@@ -5,14 +5,14 @@ namespace PasswordValidator_UnitTests
     internal class SimpleValidator_UnitTests
     {
-        SimpleValidatorFactory _factory = new SimpleValidatorFactory();
+        IValidatorFactory _factory = new PasswordValidatorFactory();
```
Same as above - do not need factory in SimpleValidator_UnitTests

IMarples reviewed (changes requested):
Plus - missing unit tests for the factory

hmarp committed:
Added PasswordValidationController Unit Tests.
Swapped PasswordValidatorFactory variable declarations in Validator unit-tests with concrete class implementations.
Added New Unit Tests for Validators.

IMarples commented on PasswordValidator/Controllers/PasswordValidationController.cs:
```diff
@@ -4,14 +4,18 @@ namespace PasswordValidator.Controllers
         public ActionResult ValidatePassword(string password)
         {
-            IPasswordValidator simpleValidator = _SimpleValidatorFactory.CreateValidator();
+            IPasswordValidator simpleValidator = _validatorFactory.GetPasswordValidator("Simple");
```
Should be ` var simpleValidator = _validatorFactory.GetPasswordValidator("Simple");`

IMarples replied:
And "Simple" should be a constant

IMarples commented on PasswordValidator/Controllers/PasswordValidationController.cs:
```diff
@@ -28,7 +32,7 @@ public ActionResult ValidatePassword(string password)
         public ActionResult ValidateAdvancedPassword(string password)
         {
-            IPasswordValidator advancedValidator = _AdvancedValidatorFactory.CreateValidator();
+            IPasswordValidator advancedValidator = _validatorFactory.GetPasswordValidator("Advanced");
```
As above var...

IMarples commented on PasswordValidator/Controllers/PasswordValidationController.cs:
```diff
@@ -28,7 +32,7 @@ public ActionResult ValidatePassword(string password)
         {
-            IPasswordValidator advancedValidator = _AdvancedValidatorFactory.CreateValidator();
+            IPasswordValidator advancedValidator = _validatorFactory.GetPasswordValidator("Advanced");
             bool isValid = advancedValidator.Validate(password);
```
var isValid..

IMarples commented on PasswordValidator/PasswordValidatorFactory.cs:
```diff
@@ -1,23 +1,22 @@
+    public interface IValidatorFactory
     {
-        public abstract IPasswordValidator CreateValidator();
+        public IPasswordValidator GetPasswordValidator(string validatorType);
```
correct

IMarples commented on PasswordValidator/PasswordValidatorFactory.cs:
```diff
@@ -1,23 +1,22 @@
-        public override IPasswordValidator CreateValidator()
-        {
-            return new AdvancedValidator();
+            if (validatorType == "Simple")
```
correct

IMarples commented on PasswordValidator_UnitTests/AdvancedValidator_UnitTests.cs:
```diff
@@ -1,22 +1,19 @@
     internal class AdvancedValidator_UnitTests
     {
-        AdvancedValidatorFactory _factory = new AdvancedValidatorFactory();
+        IPasswordValidator _validator = new AdvancedValidator();
```
still declare for the scope - private readonly and then create the instance in the constructor for the class

IMarples commented on PasswordValidator_UnitTests/AdvancedValidator_UnitTests.cs:
```diff
@@ -1,22 +1,19 @@
-
-            bool result = validator.Validate(advancedPassword);
-
+            bool result = _validator.Validate(advancedPassword);
```
var

IMarples commented on PasswordValidator_UnitTests/AdvancedValidator_UnitTests.cs:
```diff
@@ -26,11 +23,14 @@ public void IsValidAdvancedPassword_ValidPassword_ShouldReturnTrue(string advanc
-
-            bool result = validator.Validate(advancedPassword);
-
+            bool result = _validator.Validate(advancedPassword);
```
var

IMarples commented on PasswordValidator_UnitTests/AdvancedValidator_UnitTests.cs:
```diff
@@ -26,11 +23,14 @@ public void IsValidAdvancedPassword_ValidPassword_ShouldReturnTrue(string advanc
+        [TestCase(null, Description = "Null")]
+        public void IsValidAdvancedPassword_Null_ShouldThrowNullReferenceException(string advancedPassword)
+        {
+            Assert.Throws<NullReferenceException>(() => _validator.Validate(advancedPassword));
```
good

IMarples commented on PasswordValidator_UnitTests/PasswordValidationController_UnitTests.cs:
```diff
@@ -0,0 +1,67 @@
+{
+    public class PasswordValidationController_UnitTests
+    {
+        IValidatorFactory _validatorFactory = Substitute.For<IValidatorFactory>();
```
Scope them correctly and create in the constructor

IMarples commented on PasswordValidator_UnitTests/PasswordValidationController_UnitTests.cs:
```diff
@@ -0,0 +1,67 @@
+        public void ValidatePassword_ShouldReturnOkObjectResult()
+        {
+            string password = string.Empty;
+            _validatorFactory.GetPasswordValidator("Simple").Returns(_validator);
```
I like the .Returns on a new line indented

IMarples commented on PasswordValidator_UnitTests/PasswordValidationController_UnitTests.cs:
```diff
@@ -0,0 +1,67 @@
+        {
+            string password = string.Empty;
+            _validatorFactory.GetPasswordValidator("Simple").Returns(_validator);
+            _validator.Validate(password).Returns(true);
```
as above

IMarples commented on PasswordValidator_UnitTests/PasswordValidationController_UnitTests.cs:
```diff
@@ -0,0 +1,67 @@
+            string password = string.Empty;
+            _validatorFactory.GetPasswordValidator("Simple").Returns(_validator);
+            _validator.Validate(password).Returns(true);
+            PasswordValidationController controller = new PasswordValidationController(_validatorFactory);
```
var

IMarples commented on PasswordValidator_UnitTests/PasswordValidationController_UnitTests.cs:
```diff
@@ -0,0 +1,67 @@
+
+            var actionResult = controller.ValidatePassword(password);
+
+            Assert.IsInstanceOf<OkObjectResult>(actionResult);
```
good

IMarples commented on PasswordValidator_UnitTests/PasswordValidationController_UnitTests.cs:
```diff
@@ -0,0 +1,67 @@
+            _validator.Validate(password).Returns(true);
+            PasswordValidationController controller = new PasswordValidationController(_validatorFactory);
+
+            var actionResult = controller.ValidatePassword(password);
```
a var - good

IMarples commented on PasswordValidator_UnitTests/PasswordValidationController_UnitTests.cs:
```diff
@@ -0,0 +1,67 @@
+        public void ValidatePassword_ShouldReturnBadRequest()
+        {
+            string password = string.Empty;
+            _validatorFactory.GetPasswordValidator("Simple").Returns(_validator);
```
as above plus whitespace between lines 31,32 and 33

IMarples commented on PasswordValidator_UnitTests/PasswordValidationController_UnitTests.cs:
```diff
@@ -0,0 +1,67 @@
+            string password = string.Empty;
+            _validatorFactory.GetPasswordValidator("Simple").Returns(_validator);
+            _validator.Validate(password).Returns(false);
+            PasswordValidationController controller = new PasswordValidationController(_validatorFactory);
```
var and whitespace between 33 & 34

IMarples commented on PasswordValidator_UnitTests/PasswordValidationController_UnitTests.cs:
```diff
@@ -0,0 +1,67 @@
+        [Test]
+        public void ValidateAdvancedPassword_ShouldReturnOkObjectResult()
+        {
+            string password = string.Empty;
```
As above

IMarples commented on PasswordValidator_UnitTests/PasswordValidationController_UnitTests.cs:
```diff
@@ -0,0 +1,67 @@
+
+            var actionResult = controller.ValidateAdvancedPassword(password);
+
+            Assert.IsInstanceOf<BadRequestObjectResult>(actionResult);
```
good but same comments as above

IMarples commented on PasswordValidator_UnitTests/PasswordValidationController_UnitTests.cs:
```diff
@@ -0,0 +1,67 @@
+
+            Assert.IsInstanceOf<BadRequestObjectResult>(actionResult);
+        }
+    }
```
Can you put a test in to throw an exception from the mock object to raise an exception and return a 500 status code?

IMarples commented on PasswordValidator_UnitTests/SimpleValidator_UnitTests.cs:
```diff
@@ -1,21 +1,19 @@
 namespace PasswordValidator_UnitTests
 {
     internal class SimpleValidator_UnitTests
     {
```
Same as comments on previous validator tests

hmarp committed:
cleaning uP of PasswordValidationController.
Cleaning uP of unit-tests
Added test for checking Controller returns status code 500 when excPtion thrown -> FAILING
Added PasswordValidatorFactory unit-tests.

hmarp committed:
Fixed spelling mistake caused by my keyboard being broken...

IMarples commented on PasswordValidator/Controllers/PasswordValidationController.cs:
```diff
@@ -4,15 +4,28 @@ namespace PasswordValidator.Controllers
     {
-        readonly SimpleValidatorFactory _SimpleValidatorFactory = new SimpleValidatorFactory();
-        readonly AdvancedValidatorFactory _AdvancedValidatorFactory = new AdvancedValidatorFactory();
+        readonly IValidatorFactory _validatorFactory;
```
Should be `private readonly IValidatorFactory _validatorFactory;` - that way it is only 'visible/usable' from within the class and can only be created in the class constructor

IMarples commented on PasswordValidator/Controllers/PasswordValidationController.cs:
```diff
@@ -4,15 +4,28 @@ namespace PasswordValidator.Controllers
+                return StatusCode(500);
+            }
 
             if (isValid)
```
this if..else needs to be inside the `try` above

hmarp replied:
Why?

IMarples commented on PasswordValidator/Controllers/PasswordValidationController.cs:
```diff
@@ -28,8 +41,17 @@ public ActionResult ValidatePassword(string password)
+                return StatusCode(500);
+            }
 
             if (isValid)
```
Same - if..else needs to be in the `try` block

IMarples commented on PasswordValidator_UnitTests/AdvancedValidator_UnitTests.cs:
```diff
@@ -26,11 +28,14 @@ public void IsValidAdvancedPassword_ValidPassword_ShouldReturnTrue(string advanc
         [TestCase("2300fmnenvieu", Description = "No Special Character")]
```
Need more negative test cases - i.e. null, string.Empty, whitespace, other fail scenarios

IMarples commented on PasswordValidator_UnitTests/AdvancedValidator_UnitTests.cs:
```diff
@@ -1,22 +1,24 @@
 
         [TestCase("aieinr343_3@")]
         [TestCase("111dfjiea@e3")]
         [TestCase("3jda;af4fdaffede4")]
```
More positive test cases needed

IMarples commented on PasswordValidator_UnitTests/SimpleValidator_UnitTests.cs:
```diff
@@ -1,21 +1,24 @@
 
         [TestCase("0erknn3#1")]
         [TestCase("@fjeiapfio3")]
         [TestCase("fdnke4442-2")]
```
More positive test cases

IMarples commented on PasswordValidator_UnitTests/SimpleValidator_UnitTests.cs:
```diff
@@ -25,12 +28,14 @@ public void Validate_ValidPassword_ShouldReturnTrue(string password)
         [TestCase("nfn2232ier", Description = "Does Not Contain Special Character")]
```
More negative test cases
